README
srt2R
=====

Introduction
------------

Srt2R est développé par le laboratoire Topaza Pyra au sein de l'université Lyon 2.

Le logiciel permet d'exporter des fichiers de taggage au format srt vers R. 
Un fichier csv et un fichier de syntaxe R sont alors créés, permettant d'exploiter les données dans un logiciel de traitement statistique.

Création d'un paquet deb à partir des sources
---------------------------------------------

Pour créer un fichier deb, les paquets suivants doivent être présents sur la machine:
  * python-stdeb
  * fakeroot
  * python-all
  
Placez-vous ensuite dans le répertoire contenant les sources de Srt2R et lancez la commande :

python setup.py --command-packages=stdeb.command bdist_deb

Un dossier "deb_dist" est créé dans le répertoire contenant le paquet deb.

Installation
------------

Les paquets suivants doivent être présents sur votre machine :
  * python
  * python-qt4

Une fois installé avec votre gestionnaire de paquets, le logiciel apparait dans la catégorie "Accessoires" ou "Divers".

Utilisation
-----------

Pour fonctionner correctement, vos fichiers de taggage doivent respecter une certaine structure.

Srt2R supporte 3 niveaux d'étiquetage :
- niveau 1 : normalisation des noms de fichier permettant de définir des variables communes à l'ensemble des fichiers de taggage.
- niveau 2 : méta-données valables pour l'ensemble des séquences d'un fichier.
- niveau 3 : mots-clés propres à chaque séquence définie dans le fichier srt.

Les niveaux 1 et 3 sont obligatoires. La mise en place de méta-données est facultative. 

  * Étiquettes de premier niveau (convention de nommage) : 

Les variables sont placées directement dans le titre, séparées par un double underscore (__). Ex : P__01__A__20131005.srt
Un fichier "config" doit être placé dans le répertoire des fichiers srt, contenant le nom de chaque variable de titre. Les noms doivent être écrits sur une même ligne, séparés par une tabulation. Ex : Statut  Id  Version Date
 
  * Étiquettes de second niveau (méta-données) :

Les méta-données doivent être placées dans la première séquence définie dans votre fichier. 
Chaque étiquette est encadrée par des doubles astérisques (**) et contient le titre de la donnée suivie de son contenu, séparés par le signe "deux point" (:). Ex : **utilisateur:jdupont** 

  * Étiquettes de troisième niveau (tags) :

Pour chaque séquence définie dans le fichier srt, un ou plusieurs tags peuvent être associés.
Chaque tag doit être placé entre crochets. Ex : [tag] 


  * Exemple de fichier srt :

Titre : P__2__1__A__20131002.srt

Contenu :

1
00:00:00,000 --> 00:01:14,681
**intervenant:f.durand** **utilisateur:j.dupont**

2
00:02:35,173 --> 00:03:23,760
[autre]

3
00:03:23,760 --> 00:04:04,908
[tag]

4
00:04:04,908 --> 00:05:16,398
[tag2][test]

5
00:05:16,398 --> 00:05:26,335
[autre][tag]

  * Exemple de fichier config :

Contenu :

Statut	Établissement	Id	Version Date


CHANGELOG
= Release 0.4 =

   * Converts multiple srt files

   * Checks the number of variables in the filename and config file

   * Fix bug with specials characters

   * Code optimization

= Release 0.3 =

   * Retrieves metadata from first sequence
   
   * Code optimization
   
   * Add changelog and copyright files
   
   * Change section to 'misc'
   
   * Add window icon
   
= Release 0.2 =

   * Retrieves variables from filename
   
   * Retrieves filename variables names from config file
   
= Release 0.1 =

   * Initial release


srt2R/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import copy, re, ntpath, sys, linecache,  os,  codecs

from collections import OrderedDict
from PyQt4 import QtGui, QtCore
from mainwindow import Ui_MainWindow

class Srt2RApplication(QtGui.QMainWindow, Ui_MainWindow):
    def __init__(self, parent=None):
        super (Srt2RApplication, self).__init__(parent)
        self.setupUi(self)
        self.connectActions()
    
    
    def connectActions(self):
        self.srtFilesFolderButton.clicked.connect(self.openSrtFiles)
        self.folderButton.clicked.connect(self.openFolder)
        self.convertButton.clicked.connect(self.convert)
    

    def openSrtFiles(self):
        srtFilesFolder = QtGui.QFileDialog.getExistingDirectory(self,  u"Répertoire",  QtCore.QDir.homePath())
        
        if srtFilesFolder:
            self.srtFilesFolderEdit.setText(srtFilesFolder)
    
    
    def openFolder(self):
        folder = QtGui.QFileDialog.getExistingDirectory(self,  u"Répertoire",  QtCore.QDir.homePath())
        
        if folder:
            self.folderEdit.setText(folder)
    
    
    def convert(self):
        srtFilesFolder = unicode(self.srtFilesFolderEdit.text())
        folder = unicode(self.folderEdit.text())
    
    
        def getConfig(srtFilesFolder):
            
            try:
                # Retrieve title variable names from config file
                
                f = codecs.open(srtFilesFolder + '//config',  'r',  'utf-8')
                config = f.read()
                
                titleFields = config.split('\t')
                
                return titleFields
                
            except IOError,  configError:
                popupMsg(u'Erreur lors de la lecture du fichier de configuration.')
        
        
        def listSrtFiles(srtFilesFolder):
            
            try:
                srtFiles = []
            
                for fileInFolder in os.listdir(srtFilesFolder):
                        if fileInFolder.endswith('.srt'):
                            srtFiles.append(srtFilesFolder + '//' + fileInFolder)
                            
                if (srtFiles):
                    return srtFiles
                
                else:
                    popupMsg(u"Erreur : il ne semble pas y avoir de fichier srt dans ce répertoire.")
            
            except OSError,  srtFolderError:
                popupMsg(u'Erreur lors de la récupération des fichiers srt.\nLe chemin d\'accès est peut-être invalide.')
     
     
        def parser(srtFile, config):
            
            parsedFile = {}
            
            parsedFile['fileName'] = ntpath.basename(srtFile)
            
            content = file(srtFile).read()
            content = content.replace('\n',  ' ')
            
            parsedFile['lines'] = []
            titleData = []
            
            # Retrieve variable values from file title
            
            regName = '(.*?)(__|.srt)'
            titleVariables = re.findall(regName, parsedFile['fileName'])
            
            for variable in titleVariables:
                titleData.append(variable[0])
                
            if (len(titleData) == len(config)):
            
                # Retrieve metadata from first sequence
                
                firstSequence = linecache.getline(srtFile,  3)
                metadata = metaSplit(firstSequence)
                
                # First split based on the sequence number and sequence time
                
                reg = '([0-9]{1,4}\s(?:[0-9]{2}:){2}[0-9]{2},[0-9]{3} --> (?:[0-9]{2}:){2}[0-9]{2},[0-9]{3})'
                sections = re.split(reg,  content)
                del sections[0] # remove empty first line
                
                # Second split for storing elements in a double dimension array
                
                for section in sections:
                    reg2 = '[0-9]{1,4}\s((?:[0-9]{2}:){2}[0-9]{2},[0-9]{3}) --> ((?:[0-9]{2}:){2}[0-9]{2},[0-9]{3})'
                    matches = re.search(reg2,  section)
                    if (matches) :
                        debut = timeConverter(matches.group(1))
                        fin = timeConverter(matches.group(2))
                        
                        infoSequence = [(u'début', str(debut)), (u'fin', str(fin))]
                        
                        duration = fin - debut
                        infoSequence.append((u'durée', str(duration)))
                        
                        infoSequence.append((u'vidéo', parsedFile['fileName']))
                        
                        for meta in metadata:
                            infoSequence.append((meta, metadata[meta]))

                        i = 0

                        for Data in titleData :
                            infoSequence.append((config[i], Data))
                            i = i + 1
                    else :
                        tags = tagsSplit(section)
                        for tag in tags:
                            seq = copy.copy(infoSequence)
                            seq.insert(0,('tag', tag))
                            parsedFile['lines'].append(OrderedDict(seq))
                
                # Retrieve headers form first line
                
                headers = []
                
                for element in parsedFile['lines'][0]:
                    headers.append(element.strip())

                parsedFile['headers'] = headers
                
                return parsedFile
            
            else:
                popupMsg(u'Le nombre de variables contenues dans le titre de ' + parsedFile['fileName'] +' ne correspond pas avec le fichier de configuration.\nLes fichiers R n\'ont pas pu être créés pour cet élément.')


        def createCsv(fileName, folder, headersList, lines):

            csvFile = codecs.open(folder+'//'+ fileName + '.csv',  'w',  'utf-8')
            
            headers = ''
            
            for element in headersList:
                headers = headers + '\"' + element + '\",'
            
            csvFile.write(headers[:-1] + '\n')
            
            for line in lines:
                output = ''
                
                for element in line:
                    output = output + '\"' + line[element] + '\",'
                
                output = output[:-1] + '\n'
                csvFile.write(output)
                
            return True


        def timeConverter(time):
            reg = '([0-9]{2}):([0-9]{2}):([0-9]{2},[0-9]{3})'
            matches = re.search(reg,  time)
            
            hours = int(matches.group(1)) * 3600
            minutes = int(matches.group(2)) * 60
            seconds = float(re.sub(',', '.',  matches.group(3)))
            
            total = hours + minutes + seconds
            return total


        def tagsSplit(string):
            reg = '\[(.*?)\]'
            tags = re.findall(reg, string)
            
            return tags


        def metaSplit(string):
            reg = '\*\*(.*?):(.*?)\*\*'
            metadata = re.findall(reg, string)
            
            return OrderedDict(metadata)


        def createRSyntax(fileName, folder, headersList):
            rSyntaxFile = codecs.open(folder+'//'+fileName+'.R',  'w',  'utf-8')
            
            rSyntaxFile.write('if(!exists(\"Data\")) { Data <- data.frame() }\n\n')
            rSyntaxFile.write(fileName+' <- read.csv(\"'+fileName+'.csv\", na.strings = \"NA\", nrows = -1, skip = 0, check.names = TRUE, strip.white = FALSE, blank.lines.skip = TRUE)\n\n')
            
            i = 1
            
            for header in headersList:
                rSyntaxFile.write('# ' + header + '\nattributes('+fileName+')$variable.labels[' + str(i) + '] <- \"' + header + '\"\n\n')
                i = i + 1
            
            rSyntaxFile.write('Data = merge(Data, '+fileName+', all=TRUE)\n\n')
            rSyntaxFile.write('for (i in 1:length (attributes('+fileName+')$variable.labels)) { rk.set.label (Data[[i]], attributes('+fileName+')$variable.labels[i]) }\n\n')
            rSyntaxFile.write('rm('+fileName+')')
            
            return True
            
        def popupMsg(msg,  status = u'Erreur'):
            popup = QtGui.QMessageBox()
            popup.setWindowTitle(status)
            popup.setText(msg)
            popup.exec_()


        if srtFilesFolder and folder:
            
            srtFiles = listSrtFiles(srtFilesFolder)
            
            if (srtFiles):
                config = getConfig(srtFilesFolder)
                
                if (config):
                    
                    if (os.path.isdir(folder)):
                    
                        try:
                            for srtFile in srtFiles:
                                parsedFile = parser(srtFile, config)
                                
                                if (parsedFile):
                                    if (createCsv(parsedFile['fileName'],  folder,  parsedFile['headers'],  parsedFile['lines'])):
                                        createRSyntax(parsedFile['fileName'],  folder, parsedFile['headers'])
                            
                            popupMsg(u'Fichiers R créés avec succès !',  u'Succès')
                        
                        except  IOError,  e:
                            popupMsg(str(e))
                           
                    else:
                        popupMsg(u'Erreur : le répertoire de destination ne semble pas exister.')

        else :
            popupMsg(u'Vous devez spécifier un dossier contenant les fichiers srt et un dossier de destination.')


srt2R/mainwindow.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'mainwindow.ui'
#
# Created: Thu Oct  3 17:15:05 2013
#      by: PyQt4 UI code generator 4.10
#
# WARNING! All changes made in this file will be lost!

from PyQt4 import QtCore, QtGui

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtGui.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig)

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName(_fromUtf8("MainWindow"))
        MainWindow.resize(517, 221)
        icon = QtGui.QIcon()
        icon.addPixmap(QtGui.QPixmap(_fromUtf8("/usr/share/srt2R/srt2R.svg")), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        MainWindow.setWindowIcon(icon)
        self.centralwidget = QtGui.QWidget(MainWindow)
        self.centralwidget.setObjectName(_fromUtf8("centralwidget"))
        self.srtFilesFolderLabel = QtGui.QLabel(self.centralwidget)
        self.srtFilesFolderLabel.setGeometry(QtCore.QRect(40, 50, 81, 16))
        self.srtFilesFolderLabel.setObjectName(_fromUtf8("srtFilesFolderLabel"))
        self.folderLabel = QtGui.QLabel(self.centralwidget)
        self.folderLabel.setGeometry(QtCore.QRect(40, 100, 171, 16))
        self.folderLabel.setObjectName(_fromUtf8("folderLabel"))
        self.convertButton = QtGui.QPushButton(self.centralwidget)
        self.convertButton.setGeometry(QtCore.QRect(360, 150, 101, 31))
        self.convertButton.setObjectName(_fromUtf8("convertButton"))
        self.srtFilesFolderEdit = QtGui.QLineEdit(self.centralwidget)
        self.srtFilesFolderEdit.setGeometry(QtCore.QRect(120, 45, 311, 24))
        self.srtFilesFolderEdit.setObjectName(_fromUtf8("srtFilesFolderEdit"))
        self.folderEdit = QtGui.QLineEdit(self.centralwidget)
        self.folderEdit.setGeometry(QtCore.QRect(220, 95, 211, 24))
        self.folderEdit.setObjectName(_fromUtf8("folderEdit"))
        self.srtFilesFolderButton = QtGui.QToolButton(self.centralwidget)
        self.srtFilesFolderButton.setGeometry(QtCore.QRect(440, 45, 26, 22))
        self.srtFilesFolderButton.setObjectName(_fromUtf8("srtFilesFolderButton"))
        self.folderButton = QtGui.QToolButton(self.centralwidget)
        self.folderButton.setGeometry(QtCore.QRect(440, 95, 26, 22))
        self.folderButton.setObjectName(_fromUtf8("folderButton"))
        MainWindow.setCentralWidget(self.centralwidget)
        self.actionQuitter = QtGui.QAction(MainWindow)
        self.actionQuitter.setObjectName(_fromUtf8("actionQuitter"))
        self.action_propos = QtGui.QAction(MainWindow)
        self.action_propos.setObjectName(_fromUtf8("action_propos"))

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        MainWindow.setWindowTitle(_translate("MainWindow", "Convertisseur srt vers R", None))
        self.srtFilesFolderLabel.setText(_translate("MainWindow", "Fichiers srt :", None))
        self.folderLabel.setText(_translate("MainWindow", "Répertoire de destination :", None))
        self.convertButton.setText(_translate("MainWindow", "Convertir", None))
        self.srtFilesFolderButton.setText(_translate("MainWindow", "...", None))
        self.folderButton.setText(_translate("MainWindow", "...", None))
        self.actionQuitter.setText(_translate("MainWindow", "Quitter", None))
        self.action_propos.setText(_translate("MainWindow", "À propos", None))

